### DISCUSSION TITLE
Dynamic Input Ports

### DISCUSSION THREAD

**Discussion by ljilekor** [idling-mind/flowfunc#14] [Q&A]
Is it possible to create ports on the fly?

A simple, fictive example could be to create a node 'concat string' node.

It takes an undefined amount of strings as Inputs and concatenates them.
- When created, the node starts off with 1 str input
- Each time a str is entered (through connection or input port), a new input Port gets created on the fly.
- a Port is removed, when  clicking a cross or trash icon)
- allowing us to enter a 2nd str, and a 3rd, ...
- etc-
- a Port is removed, when  clicking a cross or trash icon

A bit like the MultipleSelect Control but each item gets its dedicated Port

Is this somehow possible with flowfunc? 
If so, would we need dash callbacks or is this built into flowfunc?

This would greatly enhance potential usecases for flowfunc !

**Comment by idling-mind** [comment]
It's defenitely a good idea. However, currently it's not possible using flowfunc.

Actually, there is an [example](https://flume.dev/docs/dynamic-nodes) in the js library's (flume) list of examples. 

Considering the current API of flume and dash, it will be a lot of work to implement such a feature in flowfunc. Mainly because, flume does not have that many events that I can tap into and dash also doesnt make it easy.

Another way to solve your problem (concatenating multiple strings) could be solved if one port can accept multiple connections. May be once chrisjpatty/flume#62 is fixed, this could be implemented in flowfunc.

**Comment by idling-mind** [comment]
I have added a new feature where you can define a javascript function in assets/*.js and reference it as a input to the `Node.inputs`. Take a look at the example in `examples/dynamic.py`. In the example, there are two nodes that will have a dynamic behavior; `Dynamic Ports` and `Auto Increasing Lists`. A video of `Auto Increasing Lists` is shown below. I guess this answers your current question. `Dynamic Ports` is the example from https://flume.dev/docs/dynamic-nodes


https://github.com/idling-mind/flowfunc/assets/13498682/15fab595-c445-4a39-85fc-9472a59e7a29

**Reply by ljilekor** [reply]
Greatness!!!

What about removing elements from the list?

An elegant solution could be to remove the item when disconnecting and reordering the list accordingly.
But users may use index specific elements down the graph. Maybe an auto on/off property could help.
x or trash buttons...

Nice to see the progress. I'll test it for sure.

**Reply by idling-mind** [reply]
As of now, the way this dynamic port behaves is, it counts the number of connections and creates as many plus one `str` port.
The function that makes it possible defined in `assets/funcs.js`. This is not really part of flowfunc, but it's defined alongside the dash application. So you as the developer have the freedom to redefine it.
Below is the function as it looks now.
```javascript
increasing_ports: function (ports, inputData, connections, context) {
    const arr = [];
    var connection_count = Object.keys(connections.inputs).length;
    for (let i = 0; i <= connection_count; i++) {
        arr.push(ports.str({ name: `port${i}`, label: `String ${i}` }));
    }
    return arr
}
```
Here, `ports` is the object containing all the different ports available as part of the config. `inputData` is the data defined using the controls in the node, `connections` are the connections coming into the node and context is the `context` variable that you pass in into `Flowfunc` object.
I wrote this function just as a test to see if there is possibility for the developer to define a dynamic port.
Since i'm still working on solving custom controls, adding a trash button is not yet possible.

In this particular case, considering the limitations, another solution would be to give an `int` control at the top to define how many `str` ports are needed.

Something like this

```javascript
increasing_ports: function (ports, inputData, connections, context) {
    const arr = [ports.int({name:`port_count`, label: "Number of ports needed", hidePort: true})];
    var port_count = inputData.port_count ? Object.values(inputData.port_count)[0] : 0
    // inputData.port_count is an object with keys=control names and
    // values=values of each control. There is only one control in int port
    // so taking the first value
    for (let i = 0; i < port_count; i++) {
        arr.push(ports.str({ name: `port${i}`, label: `String ${i}` }));
    }
    return arr
}
```
It behaves like this.

https://github.com/idling-mind/flowfunc/assets/13498682/11feebca-6fd8-45fc-a840-5cef9b9695e4

But again, you have the freedom to define this function. albiet in javascript and not python! 😒
